doc/developer/design/20230210_storage_persist_sink_redux.md

```diff
@@ -0,0 +1,161 @@
+# Storage: next-gen `persist_sink`
+
+Created: February 10, 2023 10:36 AM
+Status: Draft
+Tags: A-STORAGE
+
+# Overview
+
+This design document describes how the _storage_ `persist_sink` implementation should work going forward,
+in particular, how it can be upgraded to utilize all the timely workers available in a replica.
+This design is specifically chosen to consider the constraints we have for storage sources.
+
+## Goals
+
+- Unblock the "Multipurpose Clusters" epic, by effectively utilizing large replicas
+when writing source data to persist.
+  - <https://github.com/MaterializeInc/materialize/issues/17413>
+
+## Non-goals
+
+- Avoid copying code by merging portions of the compute and storage `persist_sink`'s when doing so is too onerous.
+
+# Details
+
+## Current state
+
+Currently, there are 2 implementations for sinking a differential `Collection` to persist, both called `persist_sink`.
+
+The compute implementation, described here: <https://www.notion.so/materialize/distributed-self-correcting-persist_sink-d3d59834ed9d47d397143c738e9d6c9d>
+is used to persist `MATERIALIZED VIEW`. It writes data from all available timely workers, and is _self-correcting_, which means:
+_it adjusts what the persist shard accumulates to if the persisted `Collection` has changed_. That is, if the
+`Collection` produced by a compute dataflow is somehow _different_ in a different version of the code, this sink ensures that
+the data in the persist shard is adjusted to reflect this. The compute `persist_sink` also expects to be run
+from many cluster replicas, and therefore gracefully handles contention, at the cost of write amplification.
+
+The storage implementation is much simpler. It never attempts to correct the underlying persist shard, instead only
+writing data past that shard's upper, and only writes data from a single timely worker thread. This implementation
+also irrecoverably fails if it detects contention writing to the persist shard.
+
+## Proposal
+
+Replace the storage `persist_sink` implementation with an implementation that maximizes throughput, by writing data
+from all available timely workers. This implementation will be _similar to the compute one_, but will be
+customized for the constraints storage has. These constraints are:
+
+- Uses bounded-memory.
+  - Storage `Collection`s can produce large amounts of data, over time, or within a single timestamp. We want to be
+  able to persist `Collection`s of arbitrary sizes, without needing to vertically scale.
+- Minimal overhead.
+  - Persisting a storage `Collection` should be bounded only be persist's write throughput.
+- Correctness when upstream sources compact data.
+- Required to retry on contention.
+  - Currently this means panicking, shutting down, and restarting, but in the future could be made more graceful.
+  - Compute assumes contention meant the some other replica won the `compare_and_append`.
```

Comment:
Can you say more about what goes wrong if we use the compute implementation here? Compute assumes that contention meant some other replica won .. which seems fine! I think this is maybe missing a description of how compute's retry policy would be problematic for storage.